docs/architecture/adr-059-evidence-composition-and-lifecycle.md

```diff
@@ -0,0 +1,274 @@
+
+```go
+type EvidenceInfo struct {
+  ev Evidence
+  time time.Time
+  validators []Validator
+  totalVotingPower int64
+}
+```
+
+`time`, `validators` and `totalVotingPower` are need to form the ABCI Evidence that we will come to later.
+
+
+#### Broadcasting and receiving evidence
+
+The evidence pool also runs a reactor that broadcasts the newly validated
+evidence to all connected peers.
+
+Receiving evidence from other evidence reactors works in the same manner as receiving evidence from the consensus engine or a light client.
+
+
+#### Proposing evidence on the block
+
+When it comes to prevoting and precomitting a proposal that contains evidence, the full node will once again
+call upon the evidence pool to verify the evidence using `CheckEvidence(ev []Evidence)`:
+
+This performs the following actions:
+
+1. Loops through all the evidence to check that nothing has been duplicated
+
+2. For each evidence, run `fastCheck(ev evidence)` which works similar to `Has` but instead for `LightClientAttackEvidence` if it has the
+same hash it then goes on to check that the validators it has are all signers in the commit of the conflicting header. If it doesn't pass fast check (because it hasn't seen the evidence before) then it will have to verify the evidence.
+
+3. runs `Verify(ev Evidence)` - Note: this also saves the evidence to the db as mentioned before.
+
+
+#### Updating application and pool
+
+The final part of the lifecycle is when the block is committed and the `BlockExecutor` tries to apply state. Instead of trying to form the information necessary to produce the ABCI evidence itself, it gets the evidence pool to create it when it calls `Update(Block, State) []abci.Evidence`
+
+Here is the excerpt in the `ApplyBlock()` where the evidence pool is updated.
+```go
+abciResponses.BeginBlock.ByzantineValidators = evpool.Update(block, state)
+```
+
+The `Update()` function does the following:
+
+- Increments state which keeps track of both the current time and height used for measuring expiry
+
+- Marks evidence as committed and saves to db. This prevents validators from proposing committed evidence in the future
+  Note: the db just saves the height and the hash. There is no need to save the entire committed evidence
+
+- Forms ABCI evidence:  (note for `DuplicateVoteEvidence` the validators array size is 1)
+  ```go
+  for _, val := range evInfo.Validators {
+    abciEv = append(abciEv, &abci.Evidence{
+      Type: evType,   // either DuplicateVote or LightClientAttack
+      Validator: val,   // the offending validator (which includes the address, pubkey and power)
+      Height: evInfo.ev.Height(),    // the height when the offense happened
+      Time: evInfo.time,      // the time when the offense happened
+      TotalVotingPower: evInfo.totalVotingPower   // the total voting power of the validator set
+    })
+  }
+  ```
```

Comment:
OK, so just to check my own understanding:

This snippet constructs a slice of `abci.Evidence`s. If the evidence passed in was `DuplicateVoteEvidence`, then it will be of length one. Otherwise, it will be of length _n_, where _n_ is the number of misbehaving validators.

Reply:
yes exactly